### DISCUSSION TITLE
Upgrading to V4, using email provider and all sorts of issues

### DISCUSSION THREAD

**Discussion by onefineplayandy** [nextauthjs/next-auth#5671] [Help]
### Question 💬

Hey there, 

I dont really know where to start tbh. I've just spent the last 2/3 days trying to upgrade from NextAuth v3 to v4 and i'm basically stuck and it doesnt work. 

What doesnt work you ask? I'll try and flesh that out and at this point i cant really pin point, other than it appears getting the session is the issue. 

A bit about my setup:

1. I have a custom DB adapter, heavily influenced by legacy typeORM. (this appears to work when looking at database)
2. I am using email (magic links) 
3. When i "signin" i get the entries in the database
4. when i then use the generated link with the various tokens nothing happens and i do not get redirected to my 'new user callback' instead i end up on my login screen (because all 'getSession' checks fail) 
5. Cookies are created and seem to be legit

I have painstakingly upgraded the database (the documentation was not ideal for that) i have changed all my types, models and have changed all my includes from 'next-auth/client' to 'next-auth/react'

my options are below:

```ts
export const nextAuthOptions: NextAuthOptions = {
    pages: {
        signIn: `/`,
        signOut: `/auth/logout`,
        verifyRequest: `/check`,
        newUser: `/user/new?src=na`,
        error: `/auth/error`,
    },
    // These providers will be rendered when you call /api/auth/signin
    providers: [
        EmailProvider({
            server: {
                host: `smtp.sendgrid.net`,
                port: 465,
                auth: {
                    user: process.env.SENDGRID_EMAIL_USER,
                    pass: process.env.SENDGRID_EMAIL_KEY,
                },
            },
            from: `noreply@${process.env.NEXT_PUBLIC_PRODUCT_DOMAIN_EMAIL}`,
            // Overwrite the way emails are sent from next auth and ensure we go through
            // the web api of sendgrid and the correct template is used.
            sendVerificationRequest: async ({
                identifier: email,
                url,
                token,
                provider,
            }) => {
                await nextAuthVerificationEmail({
                    email,
                    url,
                    token,
                    provider,
                });
            },
        }),
    ],
    secret: process.env.ENCRYPTION_SECRET,
    adapter: NextAuthKnexAdapter(),
    debug: true,
    callbacks: {
        /**
         * @param  {object} session      Session object
         * @param  {object} user         User object    (if using database sessions)
         *                               JSON Web Token (if not using database sessions)
         * @return {object}              Session that will be returned to the client
         */
        async session({ session, user }) {
            console.log(`callback the session`);
            // session.user.id = user.id;
            return session;
        },
    },
};

// eslint-disable-next-line import/no-anonymous-default-export
export default NextAuth(nextAuthOptions);

```

V3 worked wonderfully and TBH the only reason i'm changing it is because i deploy on Vercel with branches and i needed to have NEXTAUTH_URL => VERCEL_URL and this doesnt work anymore and V4 apparently uses VERCEL_URL if the NEXTAUTH_URL env variable isnt set. 

Any help would be great, i'm completely drawing a blank as to what the issue can be or even is. I'm currently resorting to editing the node_modules for next-auth in the vague hope that i can find some discrepency. 

### How to reproduce ☕️

i dont know what it is that you need to replicate this issue.

### Contributing 🙌🏽

No, I am afraid I cannot help regarding this

**Comment by ThangHuuVu** [comment]
Can you provide a bit more context: 
- What is the function `nextAuthVerificationEmail` look like?
- Your adapter implementation 

Curious why it didn't work for you. 👀
